[Compute] Update for EnableIPForwarding, Priority, and Identity.

hyonholee opened the pull request:
<!-- DO NOT DELETE THIS TEMPLATE -->

## Description
<!--
Please add an informative description that covers the changes made by the pull request.

If applicable, reference the bug/issue that this pull request fixes here.
-->

---

This checklist is used to make sure that common guidelines for a pull request are followed. You can find a more complete discussion of PowerShell cmdlet best practices [here](https://msdn.microsoft.com/en-us/library/dd878270(v=vs.85).aspx).

- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md).**
- [x] **If changes were made to any cmdlet, the XML help was regenerated using the [platyPSHelp module](https://github.com/Azure/azure-powershell/blob/preview/documentation/help-generation.md).**
- [x] **If any large changes are made to a service, they are reflected in the respective [change log](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log).**

### [General Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/preview/documentation/cleaning-up-commits.md).
- [x] The pull request does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md) (unless a major version change occurs in the assembly and module).

### [Testing Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.
- [x] PowerShell scripts used in tests should do any necessary setup as part of the test or suite setup, and should not use hard-coded values for locations or existing resources.

### [Cmdlet Signature Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#cmdlet-signature-guidelines)
- [x] New cmdlets that make changes or have side effects should implement `ShouldProcess` and have `SupportShouldProcess=true` specified in the cmdlet attribute. You can find more information on `ShouldProcess` [here](https://github.com/Azure/azure-powershell/wiki/PowerShell-Cmdlet-Design-Guidelines#supportsshouldprocess).
- [x] Cmdlet specifies `OutputType` attribute if any output is produced - if the cmdlet produces no output, it should implement a `PassThru` parameter.

### [Cmdlet Parameter Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#cmdlet-parameter-guidelines)
- [x] Parameter types should not expose types from the management library - complex parameter types should be defined in the module.
- [x] Complex parameter types are discouraged - a parameter type should be simple types as often as possible. If complex types are used, they should be shallow and easily creatable from a constructor or another cmdlet.
- [x] Cmdlet parameter sets should be mutually exclusive - each parameter set must have at least one mandatory parameter not in other parameter sets.

cormacpayne commented:
@hyonholee Hey Hyonho, have we had a cmdlet design review for `Update-AzureRmVmssVM` yet?

hyonholee commented:
@cormacpayne I pulled out Update-AzureRmVmssVM.  Thanks.

cormacpayne commented on src/ResourceManager/Compute/Commands.Compute/Generated/VirtualMachineScaleSet/Config/AddAzureRmVmssNetworkInterfaceConfigurationCommand.cs:
```diff
@@ -69,6 +69,11 @@ public partial class AddAzureRmVmssNetworkInterfaceConfigurationCommand : Micros
+        [Parameter(
+            Mandatory = false,
+            ValueFromPipelineByPropertyName = true)]
+        public SwitchParameter EnableIPForwarding { get; set; }
```
@hyonholee side note for the generator: switch parameters shouldn't have the `ValueFromPipelineByPropertyName` property

cormacpayne commented on src/ResourceManager/Compute/Commands.Compute/help/Grant-AzureRmDiskAccess.md:
````diff
@@ -48,6 +48,19 @@ Accept pipeline input: False
 ```
 
+### -AsJob
+Run cmdlet in the background```yaml
````
@hyonholee please move the trailing ` ```yaml ` string to a line by itself - this is due to a bug in `platyPS`. This comment goes for all markdown files in this PR with this issue

cormacpayne commented on src/ResourceManager/Compute/Commands.Compute/Manual/NewAzureRmVmssConfigCommand.cs:
```diff
@@ -0,0 +1,38 @@
+        [Parameter(
+            Mandatory = false,
+            ValueFromPipelineByPropertyName = false)]
+        [Obsolete("This parameter is obsolete.  Please use IdentityType parameter instead.", false)]
```
@hyonholee would you mind documenting this change in the [`upcoming-breaking-changes.md`](https://github.com/Azure/azure-powershell/blob/preview/src/ResourceManager/Compute/documentation/upcoming-breaking-changes.md) file?

Also, would you mind changing the text of the warning to something like the following:

`New-AzureRmVmssConfig: The parameter AssignIdentity will be removed in an upcoming release. Please use the IdentityType parameter instead.`

markcowl commented on src/ResourceManager/Compute/Commands.Compute/Generated/VirtualMachineScaleSet/Config/NewAzureRmVmssConfigCommand.cs:
```diff
@@ -158,8 +158,18 @@ public partial class NewAzureRmVmssConfigCommand : Microsoft.Azure.Commands.Reso
         [Parameter(
             Mandatory = false,
-            ValueFromPipelineByPropertyName = false)]
-        public SwitchParameter AssignIdentity { get; set; }
```
Removing this is a breakign change.  You need to add a new parameter set for additiinal user assigned identy types.

hyonholee replied:
This parameter is kept in Manual/NewAzureRmVmssConfigCommnad.cs file.

markcowl commented on src/ResourceManager/Compute/Commands.Compute/Generated/VirtualMachineScaleSet/VirtualMachineScaleSetUpdateMethod.cs:
```diff
@@ -192,6 +191,12 @@ public override void ExecuteCmdlet()
+            ParameterSetName = "DefaultParameter",
+            Mandatory = false)]
+        [ValidateNotNullOrEmpty]
+        public string[] IdentityId { get; set; }
```
This needs to go through review.  What are the possible values here, and how can we lead the user into the correct one?

hyonholee replied:
IdentityId parameter takes an array of strings (ids), each of which is a form of "/subscriptions/<sub_id> /resourceGroups/<rg_name>/providers/Microsoft.ManagedIdentity/userAssignedIdentities/<identity_name>".

User should create this ID using ManagedIdentity API (or cmdlet)

markcowl replied:
Does this not need an IdentityType? The same parameter set issues shoudl apply to both.

hyonholee replied:
This needs IdentityType with UserIdentity or SystemIdentityUserIdentity, but a user may not give any Identity at all.   So both are optional.

markcowl commented on src/ResourceManager/Compute/Commands.Compute/VirtualMachine/Config/NewAzureVMConfigCommand.cs:
```diff
@@ -62,14 +62,19 @@ public class NewAzureVMConfigCommand : Microsoft.Azure.Commands.ResourceManager.
+        [Parameter(
+            Mandatory = false,
+            ValueFromPipelineByPropertyName = true)]
+        public string[] IdentityId { get; set; }
```
same comment.  We need to lead the suer into the correct value.

markcowl commented on src/ResourceManager/Compute/Commands.Compute/VirtualMachine/Config/NewAzureVMConfigCommand.cs:
```diff
@@ -62,14 +62,19 @@ public class NewAzureVMConfigCommand : Microsoft.Azure.Commands.ResourceManager.
         [Parameter(
             ValueFromPipelineByPropertyName = false)]
         [ValidateNotNullOrEmpty]
+        [Obsolete("This parameter is obsolete.  Use IdentityType parameter instead.", false)]
```
Unclear.  We need to discuss.

markcowl commented on src/ResourceManager/Compute/Commands.Compute/VirtualMachine/Operation/UpdateAzureVMCommand.cs:
```diff
@@ -42,13 +42,18 @@ public class UpdateAzureVMCommand : VirtualMachineActionBaseCmdlet
+        [Parameter(
+            Mandatory = false,
+            ValueFromPipelineByPropertyName = true)]
+        public string[] IdentityId { get; set; }
```
same comments

markcowl commented on src/ResourceManager/Compute/documentation/upcoming-breaking-changes.md:
```diff
@@ -25,4 +25,38 @@
\ No newline at end of file
+# Upcoming Breaking Changes
+
+## Release X.0.0 - May 2018
```
Do not add this header.  Also, we need to discuss this.  The plan was not to make this a breaking change, but to add an additional parameter set - the user should still be led into the appropriate value.

maddieclayton commented on src/ResourceManager/Compute/Commands.Compute.Test/ScenarioTests/ContainerServiceTests.ps1:
```diff
@@ -140,13 +140,14 @@ function Test-ContainerServiceUpdate
             -Count 2 `
-        | Update-AzureRmContainerService
+        | Update-AzureRmContainerService -AsJob;
+        $job | Wait-Job;
```
Can you add asserts here to ensure the update has succeeded?

maddieclayton commented on src/ResourceManager/Compute/Commands.Compute.Test/ScenarioTests/DiskRPTests.ps1:
```diff
@@ -69,17 +69,19 @@ function Test-Disk
+        $job = Grant-AzureRmDiskAccess -ResourceGroupName $rgname -DiskName $diskname -Access $access -DurationInSecond 5 -AsJob;
+        $job | Wait-Job;
+        $job = Revoke-AzureRmDiskAccess -ResourceGroupName $rgname -DiskName $diskname -AsJob;
+        $job | Wait-Job;
```
Can you add asserts for Grant-AzureRmDiskAccess and Revoke-AzureRmDiskAccess to ensure they have succeeded?

maddieclayton commented on src/ResourceManager/Compute/Commands.Compute.Test/ScenarioTests/DiskRPTests.ps1:
```diff
@@ -137,17 +139,19 @@ function Test-Snapshot
+        $job = Grant-AzureRmSnapshotAccess -ResourceGroupName $rgname -SnapshotName $snapshotname -Access $access -DurationInSecond 5 -AsJob;
+        $job | Wait-Job;
+        Revoke-AzureRmSnapshotAccess -ResourceGroupName $rgname -SnapshotName $snapshotname -AsJob;
+        $job | Wait-Job;
```
Can you add asserts for Grant-AzureRmSnapshotAccess and Revoke-AzureRmSnapshotAccess to ensure they have succeeded?

maddieclayton commented on src/ResourceManager/Compute/Commands.Compute.Test/ScenarioTests/VirtualMachineScaleSetTests.ps1:
```diff
@@ -1174,11 +1243,15 @@ function Test-VirtualMachineScaleSetRollingUpgrade
+        $job = Stop-AzureRmVmssRollingUpgrade -ResourceGroupName $rgname -VMScaleSetName $vmssName -Force -AsJob;
+        $result = $job | Wait-Job;
+        Assert-AreEqual "Failed" $result.State;
+        Assert-True { $result.Error[0].ToString().Contains("There is no ongoing Rolling Upgrade to cancel.")};
```
Is it possible to add a positive test case for Start/Stop-AzureRmVmssRollingOSUpgrade?

maddieclayton commented:
@hyonholee Can you add -AsJob/tests for Remove-AzureRmVMExtension and Set-AzureRmVmssDiskEncryptionExtension as well?

hyonholee commented:
@maddieclayton  I added AsJob test with verification of the result.

markcowl commented on src/ResourceManager/Compute/Commands.Compute.Test/Commands.Compute.Test.csproj:
```diff
@@ -64,7 +64,7 @@
       <HintPath>..\..\..\packages\Microsoft.Azure.Management.Authorization.2.0.0\lib\net40\Microsoft.Azure.Management.Authorization.dll</HintPath>
```
@hyonholee Can you please add a changelog entry in the compute changelog for these changes?  Otherwise, this looks good.

hyonholee replied:
Done.  Thanks.

markcowl commented:
Pre-emptive on-demand run here: https://azuresdkci.westus2.cloudapp.azure.com/view/PowerShell/job/powershell-demand/398/

hyonholee commented on src/ResourceManager/Compute/Commands.Compute/Generated/VirtualMachineScaleSet/Config/NewAzureRmVmssConfigCommand.cs:
```diff
@@ -158,8 +158,18 @@ public partial class NewAzureRmVmssConfigCommand : Microsoft.Azure.Commands.Reso
         [Parameter(
             Mandatory = false,
-            ValueFromPipelineByPropertyName = false)]
-        public SwitchParameter AssignIdentity { get; set; }
```
Ok.  I will update PR by adding 'ExplicitIdentityParameterSet'

hyonholee commented:
@markcowl  I added changelog, and also added 'ExplicitIdentityParameterSet' for IdentityType and IdentityId.  IdentityType is also mandatory in this parameter set.

cormacpayne commented:
On-demand: https://azuresdkci.westus2.cloudapp.azure.com/job/powershell-demand/404/
